### DISCUSSION TITLE
JSON Schema for a list of objects that must contain one of three default props

### DISCUSSION THREAD

**Discussion by Roni-Ravin** [json-schema-org/community#126] [Q&A] [answered]
Hello, I need help...
How can I prepare JSON Schema for a list of objects that must contain one of three default props (1,77,90), e.g:
1. VALID:
[
  {"provider_metadata_id": 1},
  {"provider_metadata_id": 77},
  {"provider_metadata_id": 90}
]
2. VALID:
[
  {"provider_metadata_id": 77},
  {"provider_metadata_id": 44}
]
3. VALID:
[
  {"provider_metadata_id": 7884},
  {"provider_metadata_id": 4564},
  {"provider_metadata_id": 1}
]
4. INVALID
[
  {"provider_metadata_id": 7884},
  {"provider_metadata_id": 4564},
  **{"provider_metadata_id": 60}**
]

Can someone help me, please?

**Comment by musemind** [comment, marked as answer]
maybe try this:

```
{
  "type": "array",
  "items": {
    "type": "object",
    "additionalProperties": false,
    "properties": {
      "provider_metadata_id": {
        "type": "number"
      }
    }
  },
  "contains": {
    "type": "object",
    "properties": {
      "provider_metadata_id": {
        "enum": [1, 77, 90]
      }
    }
  }
}
```

**Reply by Roni-Ravin** [reply]
> maybe try this:
> 
> ```
> {
>   "type": "array",
>   "items": {
>     "type": "object",
>     "additionalProperties": false,
>     "properties": {
>       "provider_metadata_id": {
>         "type": "number"
>       }
>     }
>   },
>   "contains": {
>     "type": "object",
>     "properties": {
>       "provider_metadata_id": {
>         "enum": [1, 77, 90]
>       }
>     }
>   }
> }
> ```

God bless you! it is working! 
Thank you so much.
where can I get a good explanation or reference to the JSON SCHEMA usage guide?

**Reply by musemind** [reply]
For me, it works well: https://json-schema.org/understanding-json-schema/

**Reply by Roni-Ravin** [reply]
> For me, it works well: https://json-schema.org/understanding-json-schema/

thank you

**Reply by Roni-Ravin** [reply]
> maybe try this:
> 
> ```
> {
>   "type": "array",
>   "items": {
>     "type": "object",
>     "additionalProperties": false,
>     "properties": {
>       "provider_metadata_id": {
>         "type": "number"
>       }
>     }
>   },
>   "contains": {
>     "type": "object",
>     "properties": {
>       "provider_metadata_id": {
>         "enum": [1, 77, 90]
>       }
>     }
>   }
> }
> ```

Hi man! How can I improve the limitation to 1,77,90 but only from the following possible options: 1,2,3,4,5,77,90,100
I mean, if you will send for example 55, you will receive rejection. 
thank you!

**Comment by Roni-Ravin** [comment]
@musemind Hi man! How can I improve the limitation to 1,77,90 but only from the following possible options: 1,2,3,4,5,77,90,100
I mean, if you will send for example 55, you will receive rejection.
thank you!

**Reply by musemind** [reply]
What is the problem? Just add it to the enum array.

**Reply by Roni-Ravin** [reply]
> What is the problem? Just add it to the enum array.
I want the JSON to contain at least one of these 1,77,90 but in addition the next list will be 1,2,3,4,5,77,90.  That is, if user sent me the numbers 77, 66 - he will get an error because the number 66 is invalid.

**Reply by Roni-Ravin** [reply]
I want the JSON to contain at least one of these 1,77,90 but in addition the next list will be 1,2,3,4,5,77,90.  That is, if he sent me the numbers 77, 66 - he will get an error because the number 66 is invalid.

**Reply by musemind** [reply]
You can solve this via tuple validation with `prefixItems` for the array instead of `items`.

https://json-schema.org/understanding-json-schema/reference/array.html#tuple-validation

**Reply by karenetheridge** [reply]
See also `contains`.

**Reply by Roni-Ravin** [reply]
Hello, first of all - thank you for your responses, I appreciate it.
I tried to do what you said - it is not working.
Please take a look at my example there: 
https://www.jsonschemavalidator.net/s/srClw0hB
This JSON:
`[
  {"provider_metadata_id": 5555},
  {"provider_metadata_id": 1}
]`
Should be incorrect for the following enum:
`        "enum": [
          1,
          77,
          90,
          12345
        ]`
I mention that all the items (provider_metadata_id) in the array must be one from the following list 12345, 1, 77, 90
And at least one of them must appear 1 or 77 or 90.
@musemind 
@karenetheridge

**Comment by karenetheridge** [comment]
In your link, you have the "required" in the wrong place. It should be inside the array definition.

Also, `prefixItems` is entirely ignored there because that implementation doesn't support draft2020-12. (In earlier drafts, `prefixItems` was called `items`.)

This link correctly shows an error: https://www.jsonschemavalidator.net/s/EAxF9E3U

**Reply by Roni-Ravin** [reply]
@karenetheridge @musemind thank you all guys! I really appreciate your help!

**Comment by Roni-Ravin** [comment]
@karenetheridge @musemind Hi :-( 
It does not work on my code...
I am using 
    "jsonschema": "^1.4.0",

this is my schema object:
```
{
  id: '/registerVehicleIdentityPayload',
  type: 'array',
  items: {
    type: 'object',
    properties: {
     provider_metadata_id: {type: 'integer', enum: [1,77,90,12345},
     value: {type: ['string', 'number', 'boolean']},
     confidence: {type: 'number'},
     key: {type: 'string'}
   },
    required: ['provider_metadata_id', 'value']
  },
  contains: {
    type: 'object',
    properties: {
      provider_metadata_id: {enum: [1, 77, 90]}
    }
  }
}
```

and the following JSON passes the schema (it should fail because it supposed to contain one of the 1,77,90):
```
[
  {"provider_metadata_id": 12345}
]
```
Please help me :-(

**Reply by karenetheridge** [reply]
Sounds like there might be a bug in the implementation you are using.  This works: https://www.jsonschemavalidator.net/s/dQKWkVR9

**Reply by Roni-Ravin** [reply]
@karenetheridge yes... but where is the bug? how is it possible? I really tried everything.....

**Reply by karenetheridge** [reply]
Software is not perfect :). I suggest you file a ticket with the implementation that you are using.
